File: backend/src/models/Trip.js
import mongoose from 'mongoose';

const tripSchema = new mongoose.Schema(
  {
    tripName: { type: String, required: true, trim: true, maxlength: 100, alias: 'title' },
    destination: { type: String, required: true, trim: true, maxlength: 100 },
    country: { type: String, required: true, trim: true, maxlength: 100 },
    city: { type: String, required: true, trim: true, maxlength: 100 },
    startDate: { type: Date, required: true },
    endDate: { type: Date, required: true },
    activities: { type: [String], default: [] },
    notes: { type: String, maxlength: 2000, default: '' },
    weatherSummary: { type: String, maxlength: 500, default: '' },
    averageTemperature: { type: Number, default: null },
    owner: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
      index: true,
      alias: 'userId',
    },
  },
  {
    timestamps: true,
    toJSON: { virtuals: true },
    toObject: { virtuals: true },
  }
);

tripSchema.index({ owner: 1, startDate: 1 });
tripSchema.index({ tripName: 'text', destination: 'text', country: 'text', city: 'text', activities: 'text' });

const Trip = mongoose.model('Trip', tripSchema);
export default Trip;

File: lib/features/trip/domain/trip.dart
class Trip {
  Trip({
    required this.id,
    required this.tripName,
    required this.destination,
    required this.country,
    required this.city,
    required this.startDate,
    required this.endDate,
    required this.activities,
    required this.notes,
  });

  final String id;
  final String tripName;
  final String destination;
  final String country;
  final String city;
  final DateTime startDate;
  final DateTime endDate;
  final List<String> activities;
  final String notes;

  factory Trip.fromJson(Map<String, dynamic> json) {
    return Trip(
      id: json['_id'] as String? ?? json['id'] as String? ?? '',
      tripName: json['tripName'] as String? ?? '',
      destination: json['destination'] as String? ?? '',
      country: json['country'] as String? ?? '',
      city: json['city'] as String? ?? '',
      startDate:
          DateTime.tryParse(json['startDate'] as String? ?? '') ??
          DateTime.now(),
      endDate:
          DateTime.tryParse(json['endDate'] as String? ?? '') ?? DateTime.now(),
      activities:
          (json['activities'] as List<dynamic>?)
              ?.map((item) => item.toString())
              .toList() ??
          [],
      notes: json['notes'] as String? ?? '',
    );
  }
}
